File: 0.TimeSpaceComplexity/binarySearch.js
const binarySearch = (arr, target) => {
    arr.sort((a, b) => a - b); // Ensure the array is sorted
    let left = 0;
    let right = arr.length - 1;

    while (left <= right) {
        const mid = Math.floor((left + right) / 2);

        if (arr[mid] === target) {
            return mid; // Target found at index mid
        } else if (arr[mid] < target) {
            left = mid + 1; // Search in the right half
        } else {
            right = mid - 1; // Search in the left half
        }
    }

    return -1; // Target not found
}

// Example usage:
const array = [1, 3, 5, 7, 9, 11, 13, 15];
const targetValue = 7;

const resultIndex = binarySearch(array, targetValue);

if (resultIndex !== -1) {
    console.log(`Target ${targetValue} found at index ${resultIndex}.`);
} else {
    console.log(`Target ${targetValue} not found in the array.`);
}

File: 0.TimeSpaceComplexity/binary_search.py
def binary_search(arr, target):
    arr = sorted(arr)  # Ensure the array is sorted
    left, right = 0, len(arr) - 1

    while left <= right:
        mid = len(arr) // 2
        if arr[mid] == target:
            return mid  # Target found at index mid
        elif arr[mid] < target:
            left = mid + 1  # Search in the right half
        else:
            right = mid - 1  # Search in the left half

    return -1  # Target not found


# Example usage:
numbers = input("Enter numbers separated by spaces: ").split()
numbers = [int(num) for num in numbers]
target = int(input("Enter the target number: "))
result = binary_search(numbers, target)

print(f"The target number {target} is at index: {result}" if result != -1 else f"The target number {target} is not in the list.")